### PR TITLE
feat(vertexai): add vertex_ai_index_endpoint for Vertex AI Matching Engine

### PR DESCRIPTION
**Pull request by shotarok**
Part of https://github.com/hashicorp/terraform-provider-google/issues/12818
Part of https://github.com/hashicorp/terraform-provider-google/issues/9298

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-the-terraform-providers), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/main/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
vertex_ai_index_endpoint
```

### CODE REVIEW

**Comment by shotarok** [comment]
@roaks3 All tests passed. When you have time, could you please revive this PR? Thanks

**Comment by shotarok** [comment]
@roaks3 The failed test is not related to the changes in this PR. So could you review this PR?  Apart from that, the failed test is what I added in the previous PR. So I'd appreciate it if I could know the error message to fix it in a separate PR :pray:

File: mmv1/products/vertexai/api.yaml [hunk 1]
```diff
@@ -806,6 +806,7 @@ objects:
             description: |
               The disk utilization of the MetadataStore in bytes.
             output: true
+
```
Comment (roaks3) [inline mmv1/products/vertexai/api.yaml:809]: Nit: While we don't enforce spacing rules in these files, this still seems extraneous as part of this PR. FWIW, I think most `api.yaml` files do not use extra newlines between resources.

File: mmv1/products/vertexai/api.yaml [hunk 2]
```diff
@@ -997,3 +998,74 @@ objects:
...
+        message: 'message'
+    description: |-
+      An endpoint indexes are deployed into. An index endpoint can have multiple deployed indexes.
+    parameters:
+      - !ruby/object:Api::Type::String
+        name: region
+        description: The region of the index endpoint. eg us-central1
+        url_param_only: true
+        input: true
+    properties:
```
Comment (roaks3) [inline mmv1/products/vertexai/api.yaml:1038]: Any reason for not including `deployedIndexes[]`? https://cloud.google.com/vertex-ai/docs/reference/rest/v1/projects.locations.indexEndpoints#DeployedIndex

**Review by roaks3 — CHANGES_REQUESTED** [review]
Everything looks good, I just wanted to check if there was a reason for leaving `deployedIndexes` out. Ideally we would probably want to include it within this changeset.

File: mmv1/products/vertexai/api.yaml (hunk 1 above)
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml:809]: Thank you for pointing it out! I realized `vertexai/api.yaml` has a white line between resources while resolving a merge conflict. Therefore, I added the white line. what do you think if I remove the other white lines in `vertexai/api.yaml`? Or would you like to remove this change from this PR?

File: mmv1/products/vertexai/api.yaml (hunk 2 above)
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml:1038]: Thank you for your review! A reason is the `deployedIndexes[]` is the output-only field. So I didn't define it here. 

Another reason is I plan to create `google_vertex_ai_deployed_index` to manage a deployed index. I'll use the `deployIndexes[]` when implementing it for the read operation. The description of https://github.com/hashicorp/terraform-provider-google/issues/12818 has the ideas behind `google_vertex_ai_deployed_index` and its corresponding API endpoints.

What do you think?

File: mmv1/products/vertexai/api.yaml (hunk 1 above)
Comment (shotarok) [reply mmv1/products/vertexai/api.yaml]: Agreed. I deleted the white line in [5d839be](https://github.com/GoogleCloudPlatform/magic-modules/pull/6873/commits/5d839be9ac55d2977e6897d2a1518ff524a8a6d3).

File: mmv1/products/vertexai/api.yaml (hunk 2 above)
Comment (roaks3) [reply mmv1/products/vertexai/api.yaml:1037]: Yep, this looks good to me, thanks!

**Review by roaks3 — APPROVED** [review]
LGTM, adding another reviewer for follow-up

**Comment by shotarok** [comment]
@roaks3 Thank you for your reviews! 

@slevenick Could you review this PR when you have time? Thanks!

**Review by slevenick — DISMISSED** [review]
Can you add an update test for this resource? We write handwritten update tests to make sure the resource can be modified in-place correctly and that the fields marked as updatable (in this case it looks like network is updatable?) are actually updatable

**Comment by shotarok** [comment]
@roaks3 @slevenick Could you help me understand the errors in the build logs? I'm afraid I can't see that link.

**Comment by roaks3** [comment]
Hey @shotarok , here's the error I'm seeing in our logs:
```
FAIL	github.com/hashicorp/terraform-provider-google/google [setup failed]
# github.com/hashicorp/terraform-provider-google/google
google/resource_vertex_ai_index_endpoint_test.go:146:3: expected ';', found mmv1
?   	github.com/hashicorp/terraform-provider-google	[no test files]
?   	github.com/hashicorp/terraform-provider-google/version	[no test files]
FAIL
make: *** [test] Error 1
```

Note that you should be able to run your tests locally, which will shorten the feedback loop and prevent access issues like this: https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/

**Comment by shotarok** [comment]
Thank you, and I apologize for my laziness 🙇 I'll push a commit after running the test locally.

**Comment by gericdong** [comment]
@shotarok: hello, I am checking if you're planning to move this PR forward. Let's know if there is any issue that we can help with. Thanks.

**Comment by shotarok** [comment]
@gericdong (CC: @roaks3) Thank you for your heads-up, and sorry for keeping this PR open for a long time. I resolved the conflict and removed unnecessary code from the unit test with the recent commits.

**Comment by shotarok** [comment]
@roaks3 Thank you for your approval to run the acceptance test. I understand I can run the failed tests locally, but it takes hours to create even a single index... so I'd appreciate it if I could see the error messages of the index endpoint tests 🙏

**Comment by roaks3** [comment]
@shotarok That's fair, but note that it makes the feedback loops much longer. Are you able to see the most recent failures with the "[Error message] [Debug log]" links?

It looks like your tests are sharing some config, and one may be altering the other before it is complete. I would make sure that the infrastructure for the tests is isolated, and in particular, the bootstrapped networks should be different. So `BootstrapSharedTestNetwork(t, "vertex")` should probably be split into `BootstrapSharedTestNetwork(t, "vertex-ai-index-endpoint")` and `BootstrapSharedTestNetwork(t, "vertex-ai-index-endpoint-update")` for the respective tests.

**Comment by shotarok** [comment]
Thank you so much for your comments! I understood them.

> Are you able to see the most recent failures with the "[Error message] [Debug log]" links?

I couldn't see them. Both links show me the 403 error like the below one. It'd be great If all contributors could see the error messages.

image

**Comment by roaks3** [comment]
Ah ok, I misunderstood that those messages are not available to everyone to prevent leaking secrets.

Here are the errors:
```
    testing_new.go:84: Error running post-test destroy, there may be dangling resources: exit status 1
        
        Error: Error when reading or editing ServiceNetworkingConnection "projects%2Fci-test-project-188019%2Fglobal%2Fnetworks%2Ftf-bootstrap-net-vertex:servicenetworking.googleapis.com": googleapi: Error 400: There is no peering (servicenetworking-googleapis-com) for the network (GLOBAL:global/PROJECT:1067888929963/NETWORK:tf-bootstrap-net-vertex)., badRequest
        
--- FAIL: TestAccVertexAIIndexEndpoint_vertexAiIndexEndpointExample (72.72s)
```
```
    vcr_utils.go:145: Step 1/4 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # google_service_networking_connection.vertex_vpc_connection will be updated in-place
          ~ resource "google_service_networking_connection" "vertex_vpc_connection" {
                id                      = "projects%2Fci-test-project-188019%2Fglobal%2Fnetworks%2Ftf-bootstrap-net-vertex:servicenetworking.googleapis.com"
              ~ reserved_peering_ranges = [
                  - "tf-test-address-namemqu7ve1buu",
                  + "tf-test-address-namexuwf9h8qj1",
                ]
                # (3 unchanged attributes hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccVertexAIIndexEndpoint_updated (90.94s)
```

**Comment by shotarok** [comment]
@roaks3 (CC:  @slevenick) I fixed unexpected label definitions in the test. Could you approve the cloud builder to trigger the acceptance test at your convenience?

**Comment by roaks3** [comment]
```
    vcr_utils.go:145: Step 3/4 error: Error running apply: exit status 1
        
        Error: Error waiting for Updating IndexEndpoint: timeout while waiting for state to become 'done: true' (last state: 'done: false', timeout: 20m0s)
        
          with google_vertex_ai_index_endpoint.index_endpoint,
          on terraform_plugin_test.tf line 2, in resource "google_vertex_ai_index_endpoint" "index_endpoint":
           2: resource "google_vertex_ai_index_endpoint" "index_endpoint" {
--- FAIL: TestAccVertexAIIndexEndpoint_updated (1277.51s)
```

Based on the timestamps in the logs, it looks like it legitimately timed out after 20 mins. Possibly needs a bump, or something is hanging.

**Comment by shotarok** [comment]
@roaks3 I increased the timeout from 20 min to 1h for the vertex ai index endpoint operations. If the change looks good, could you approve the cloud builder?

**Comment by roaks3** [comment]
No worries! Yes, it looks like we hit the timeout again (1 hr). We could try bumping it up again much higher, like 2h, and use that to diagnose if there is some other issue.

Relatedly, while we want the timeout to be high enough that users will not hit it under reasonable usage, we may additionally want to look at making this particular test faster (just so it doesn't bog down the nightly tests).

File: mmv1/products/vertexai/IndexEndpoint.yaml [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  error: !ruby/object:Api::OpAsync::Error
+    path: 'error'
+    message: 'message'
+description: |-
+  An endpoint indexes are deployed into. An index endpoint can have multiple deployed indexes.
+autogen_async: false
+timeouts: !ruby/object:Api::Timeouts
+  insert_minutes: 60
+  update_minutes: 60
+  delete_minutes: 60
```
Comment (shotarok) [inline mmv1/products/vertexai/IndexEndpoint.yaml:47]:
```suggestion
  insert_minutes: 120
  update_minutes: 120
  delete_minutes: 120
```

**Comment by shotarok** [comment]
@roaks3 Thank you for your comment! I've updated the timeouts to 2 hours. 

> we may additionally want to look at making this particular test faster 

Yeah, that'd be great. At the same time, I observed that a long-running operation to create an index endpoint took a long time. So I'm not sure how much we can make unit tests faster only changing test code...

**Comment by shotarok** [comment]
@roaks3 Could you take a look at the crash logs? Is it worth retrying the acceptance test? Or is there a related error to vertex-ai-index-endpoint?

**Comment by roaks3** [comment]
Looks like the test did timeout again after 2h. This is the log up until the timeout, and what's strange is that I'm not seeing any vertexai requests (just the network being created). I am wondering if something else is wrong here:

```
=== RUN   TestAccVertexAIIndexEndpoint_updated
=== PAUSE TestAccVertexAIIndexEndpoint_updated
=== CONT  TestAccVertexAIIndexEndpoint_updated
2023/05/24 15:27:24 [INFO] Authenticating using configured Google JSON 'credentials'...
2023/05/24 15:27:24 [INFO]   -- Scopes: [https://www.googleapis.com/auth/cloud-platform https://www.googleapis.com/auth/userinfo.email]
2023/05/24 15:27:24 [INFO] Authenticating using configured Google JSON 'credentials'...
2023/05/24 15:27:24 [INFO]   -- Scopes: [https://www.googleapis.com/auth/cloud-platform https://www.googleapis.com/auth/userinfo.email]
2023/05/24 15:27:24 [DEBUG] Waiting for state to become: [success]
2023/05/24 15:27:24 [INFO] Terraform is using this identity: hashicorp-test-runner@ci-test-project-188019.iam.gserviceaccount.com
2023/05/24 15:27:24 [DEBUG] Getting shared test network "tf-bootstrap-net-vertex-ai-index-endpoint-update"
2023/05/24 15:27:24 [INFO] Instantiating GCE client for path https://compute.googleapis.com/compute/beta/
2023/05/24 15:27:24 [DEBUG] Retry Transport: starting RoundTrip retry loop
2023/05/24 15:27:24 [DEBUG] Retry Transport: request attempt 0
2023/05/24 15:27:24 [DEBUG] Google API Request Details:
---[ REQUEST ]---------------------------------------
GET /compute/beta/projects/ci-test-project-188019/global/networks/tf-bootstrap-net-vertex-ai-index-endpoint-update?alt=json&prettyPrint=false HTTP/1.1
Host: compute.googleapis.com
User-Agent: google-api-go-client/0.5
X-Goog-Api-Client: gl-go/1.19.9 gdcl/0.121.0
Accept-Encoding: gzip


-----------------------------------------------------
2023/05/24 15:27:24 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Wed, 24 May 2023 15:27:24 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
 "kind": "compute#network",
 "id": "4533200781331544447",
 "creationTimestamp": "2023-05-19T07:28:32.452-07:00",
 "name": "tf-bootstrap-net-vertex-ai-index-endpoint-update",
 "selfLink": "https://www.googleapis.com/compute/beta/projects/ci-test-project-188019/global/networks/tf-bootstrap-net-vertex-ai-index-endpoint-update",
 "selfLinkWithId": "https://www.googleapis.com/compute/beta/projects/ci-test-project-188019/global/networks/4533200781331544447",
 "autoCreateSubnetworks": false,
 "routingConfig": {
  "routingMode": "REGIONAL"
 },
 "networkFirewallPolicyEnforcementOrder": "AFTER_CLASSIC_FIREWALL"
}
-----------------------------------------------------
2023/05/24 15:27:24 [DEBUG] Retry Transport: Stopping retries, last request was successful
2023/05/24 15:27:24 [DEBUG] Retry Transport: Returning after 1 attempts
2023/05/24 15:27:24 [INFO] Instantiating GCE client for path https://compute.googleapis.com/compute/beta/
2023/05/24 15:27:24 [DEBUG] Retry Transport: starting RoundTrip retry loop
2023/05/24 15:27:24 [DEBUG] Retry Transport: request attempt 0
2023/05/24 15:27:24 [DEBUG] Google API Request Details:
---[ REQUEST ]---------------------------------------
GET /compute/beta/projects/ci-test-project-188019/global/networks/tf-bootstrap-net-vertex-ai-index-endpoint-update?alt=json&prettyPrint=false HTTP/1.1
Host: compute.googleapis.com
User-Agent: google-api-go-client/0.5
X-Goog-Api-Client: gl-go/1.19.9 gdcl/0.121.0
Accept-Encoding: gzip


-----------------------------------------------------
2023/05/24 15:27:25 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Wed, 24 May 2023 15:27:25 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
 "kind": "compute#network",
 "id": "4533200781331544447",
 "creationTimestamp": "2023-05-19T07:28:32.452-07:00",
 "name": "tf-bootstrap-net-vertex-ai-index-endpoint-update",
 "selfLink": "https://www.googleapis.com/compute/beta/projects/ci-test-project-188019/global/networks/tf-bootstrap-net-vertex-ai-index-endpoint-update",
 "selfLinkWithId": "https://www.googleapis.com/compute/beta/projects/ci-test-project-188019/global/networks/4533200781331544447",
 "autoCreateSubnetworks": false,
 "routingConfig": {
  "routingMode": "REGIONAL"
 },
 "networkFirewallPolicyEnforcementOrder": "AFTER_CLASSIC_FIREWALL"
}
-----------------------------------------------------
2023/05/24 15:27:25 [DEBUG] Retry Transport: Stopping retries, last request was successful
2023/05/24 15:27:25 [DEBUG] Retry Transport: Returning after 1 attempts
panic: test timed out after 2h0m0s
```

**Comment by shotarok** [comment]
Thank you for putting the logs. The following tests are similar because the 1st one tests creation and the 2nd one tests update. However, the only 1st one succeeds now... I couldn't find any concern although I read Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-6873-old..auto-pr-6873), though. 

- `TestAccVertexAIIndexEndpoint_vertexAiIndexEndpointExample`
- `TestAccVertexAIIndexEndpoint_updated`

I can't see the whole error logs. So I'd appreciate it if I could get some advice.

---

BTW, Vertex AI Index also has 3 hours timeout now (https://github.com/GoogleCloudPlatform/magic-modules/pull/7657). Can we give a try to 3 hours timeout for Index Endpoint too?

**Comment by roaks3** [comment]
Possibly, although I think I'm seeing different behavior than expected locally. It looks like the PATCH call succeeds, but then we get stuck in some sort of retry loop. I can't tell if it's an operation that is returning something different than expected, or perhaps some sort of error that is being retried.

I will be able to look closer on Tuesday, but posting some of the logs here if you want to try to fix. Actually as I'm typing this, I'm wondering if the problem is simply that the `PATCH` call is not async (so it does not return an operation, and our handling thinks it needs to keep trying).

```
PATCH /v1/projects/terraform-dev-ryanoaks/locations/us-central1/indexEndpoints/2559830195229949952?alt=json&updateMask=description%2Clabels HTTP/1.1
Host: us-central1-aiplatform.googleapis.com
User-Agent: Terraform/1.4.6 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google/acc
Content-Length: 243
Content-Type: application/json
Accept-Encoding: gzip

{
 "description": "A sample vertex endpoint (updated)",
 "displayName": "sample-endpoint",
 "labels": {
  "label-one": "value-one",
  "label-two": "value-two"
 },
 "network": "projects/800453553510/global/networks/tf-bootstrap-net-vertex-ai-index-endpoint-update"
}

-----------------------------------------------------
2023/05/25 17:48:52 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Thu, 25 May 2023 21:48:52 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "name": "projects/800453553510/locations/us-central1/indexEndpoints/2559830195229949952",
  "displayName": "sample-endpoint",
  "description": "A sample vertex endpoint (updated)",
  "labels": {
    "label-one": "value-one",
    "label-two": "value-two"
  },
  "createTime": "2023-05-25T21:48:32.393920Z",
  "updateTime": "2023-05-25T21:48:52.559737Z",
  "network": "projects/800453553510/global/networks/tf-bootstrap-net-vertex-ai-index-endpoint-update"
}

-----------------------------------------------------
2023/05/25 17:48:52 [DEBUG] Retry Transport: Stopping retries, last request was successful
2023/05/25 17:48:52 [DEBUG] Retry Transport: Returning after 1 attempts
2023/05/25 17:48:52 [DEBUG] Finished updating IndexEndpoint "projects/terraform-dev-ryanoaks/locations/us-central1/indexEndpoints/2559830195229949952": map[string]interface {}{"createTime":"2023-05-25T21:48:32.393920Z", "description":"A sample vertex endpoint (updated)", "displayName":"sample-endpoint", "labels":map[string]interface {}{"label-one":"value-one", "label-two":"value-two"}, "name":"projects/800453553510/locations/us-central1/indexEndpoints/2559830195229949952", "network":"projects/800453553510/global/networks/tf-bootstrap-net-vertex-ai-index-endpoint-update", "updateTime":"2023-05-25T21:48:52.559737Z"}
2023/05/25 17:48:52 [DEBUG] Waiting for state to become: [done: true]
2023/05/25 17:48:52 [DEBUG] Waiting for state to become: [success]
2023/05/25 17:48:52 [DEBUG] Retry Transport: starting RoundTrip retry loop
2023/05/25 17:48:52 [DEBUG] Retry Transport: request attempt 0
2023/05/25 17:48:52 [DEBUG] Google API Request Details:
---[ REQUEST ]---------------------------------------
GET /v1/projects/800453553510/locations/us-central1/indexEndpoints/2559830195229949952?alt=json HTTP/1.1
Host: us-central1-aiplatform.googleapis.com
User-Agent: Terraform/1.4.6 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google/acc
Content-Type: application/json
Accept-Encoding: gzip


-----------------------------------------------------
2023/05/25 17:48:52 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Thu, 25 May 2023 21:48:52 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "name": "projects/800453553510/locations/us-central1/indexEndpoints/2559830195229949952",
  "displayName": "sample-endpoint",
  "description": "A sample vertex endpoint (updated)",
  "labels": {
    "label-one": "value-one",
    "label-two": "value-two"
  },
  "createTime": "2023-05-25T21:48:32.393920Z",
  "updateTime": "2023-05-25T21:48:52.559737Z",
  "network": "projects/800453553510/global/networks/tf-bootstrap-net-vertex-ai-index-endpoint-update"
}

-----------------------------------------------------
2023/05/25 17:48:52 [DEBUG] Retry Transport: Stopping retries, last request was successful
2023/05/25 17:48:52 [DEBUG] Retry Transport: Returning after 1 attempts
2023/05/25 17:48:52 [DEBUG] Got done: false while polling for operation projects/800453553510/locations/us-central1/indexEndpoints/2559830195229949952's status
2023/05/25 17:49:02 [DEBUG] Waiting for state to become: [success]
2023/05/25 17:49:02 [DEBUG] Retry Transport: starting RoundTrip retry loop
2023/05/25 17:49:02 [DEBUG] Retry Transport: request attempt 0
2023/05/25 17:49:02 [DEBUG] Google API Request Details:
---[ REQUEST ]---------------------------------------
GET /v1/projects/800453553510/locations/us-central1/indexEndpoints/2559830195229949952?alt=json HTTP/1.1
Host: us-central1-aiplatform.googleapis.com
User-Agent: Terraform/1.4.6 (+https://www.terraform.io) Terraform-Plugin-SDK/2.10.1 terraform-provider-google/acc
Content-Type: application/json
Accept-Encoding: gzip


-----------------------------------------------------
2023/05/25 17:49:02 [DEBUG] Google API Response Details:
---[ RESPONSE ]--------------------------------------
HTTP/2.0 200 OK
Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
Cache-Control: private
Content-Type: application/json; charset=UTF-8
Date: Thu, 25 May 2023 21:49:02 GMT
Server: ESF
Vary: Origin
Vary: X-Origin
Vary: Referer
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 0

{
  "name": "projects/800453553510/locations/us-central1/indexEndpoints/2559830195229949952",
  "displayName": "sample-endpoint",
  "description": "A sample vertex endpoint (updated)",
  "etag": "AMEw9yOdLqHjdQIcFBHD8OQWEhPOFuaaos9MvJ0DW--J82ouOign-RHP2T4KLYoQ-do3",
  "labels": {
    "label-one": "value-one",
    "label-two": "value-two"
  },
  "createTime": "2023-05-25T21:48:32.393920Z",
  "updateTime": "2023-05-25T21:48:52.559737Z",
  "network": "projects/800453553510/global/networks/tf-bootstrap-net-vertex-ai-index-endpoint-update"
}
```

**Comment by roaks3** [comment]
One last note for the day: I had this test running locally for the last 3.5 hrs and it was still stuck in this polling state.

**Comment by shotarok** [comment]
Thank you so much for your responses and notes 🙏 They're helpful for me to understand the problem

File: mmv1/products/vertexai/IndexEndpoint.yaml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+name: IndexEndpoint
+base_url: projects/{{project}}/locations/{{region}}/indexEndpoints
+create_url: projects/{{project}}/locations/{{region}}/indexEndpoints
+self_link: projects/{{project}}/locations/{{region}}/indexEndpoints/{{name}}
+update_verb: :PATCH
+update_mask: true
+create_verb: :POST
+references: !ruby/object:Api::Resource::ReferenceLinks
+  api: https://cloud.google.com/vertex-ai/docs/reference/rest/v1/projects.locations.indexEndpoints/
+async: !ruby/object:Api::OpAsync
```
Comment (roaks3) [inline mmv1/products/vertexai/IndexEndpoint.yaml:24]:
```suggestion
async: !ruby/object:Api::OpAsync
  actions:
    - create
    - delete
```

File: mmv1/products/vertexai/IndexEndpoint.yaml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    allowed:
+      - True
+      - False
+  error: !ruby/object:Api::OpAsync::Error
+    path: 'error'
+    message: 'message'
+description: |-
+  An endpoint indexes are deployed into. An index endpoint can have multiple deployed indexes.
+autogen_async: false
+timeouts: !ruby/object:Api::Timeouts
```
Comment (roaks3) [inline mmv1/products/vertexai/IndexEndpoint.yaml:44]: I think we can remove these, and rely on the 20m default

**Review by roaks3 — CHANGES_REQUESTED** [review]
Ok yea, looks like the update method is not async, so it just sits there and polls. I've added a suggestion to fix, and verified the test looked good locally (it took 90 seconds).

As an additional help, if you wouldn't mind catching this branch up with the latest changes on `main`, it should go a little faster if we need more local testing.

File: mmv1/products/vertexai/IndexEndpoint.yaml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  insert_minutes: 120
+  update_minutes: 120
+  delete_minutes: 120
```
Comment (shotarok) [inline mmv1/products/vertexai/IndexEndpoint.yaml:47]:
```suggestion
```

File: mmv1/products/vertexai/IndexEndpoint.yaml (hunk 5 above)
Comment (shotarok) [reply mmv1/products/vertexai/IndexEndpoint.yaml]: thanks, I made changes along with your suggestion in [8aeccb0](https://github.com/GoogleCloudPlatform/magic-modules/pull/6873/commits/8aeccb0602f3d80d19eede04abfc27a548662cc3).

**Comment by shotarok** [comment]
@roaks3 Sorry for the delayed response. I was taking consecutive days off. I pushed the changes based on your suggestions! BTW, `cla/google` started failing because 8aeccb0 was co-authored by both emails of us. Please let me know if you'd like me to make a force-push to remove the commit. Thank you for your support!

**Comment by shotarok** [comment]
@roaks3 It looks like the acceptance tests related to vertex_ai_index_endpoint  finally succeeded. Thank you so much for your support!

**Review by roaks3 — COMMENTED** [review]
I took another pass since it's been a while, and this largely looks good to merge. I only had two remaining comments to make sure that updates are configured correctly.

It would be ideal for you to check these configs locally, but in case that is not an option, I think it would be reasonable to submit changes that attempt to change these values in the test, and see if there are errors in CI that we need to resolve.

File: mmv1/third_party/terraform/tests/resource_vertex_ai_index_endpoint_test.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	return Nprintf(`
+resource "google_vertex_ai_index_endpoint" "index_endpoint" {
+  display_name = "sample-endpoint"
+  description  = "A sample vertex endpoint (updated)"
+  region       = "us-central1"
+  labels       = {
+    label-one = "value-one"
+    label-two = "value-two"
+  }
+  network      = "projects/${data.google_project.project.number}/global/networks/${data.google_compute_network.vertex_network.name}"
```
[the remark this answers is not in the document]
Comment (shotarok) [reply mmv1/third_party/terraform/tests/resource_vertex_ai_index_endpoint_test.go:89]: When I run the unit test locally, I faced the following error. I don't think we can replace the network based on the error message. I'll make it immutable. Thanks for catching it up!

```
 Error: Error creating IndexEndpoint: googleapi: Error 400: Currently only one VPC network per user project is supported. Your project is already linked to https://www.googleapis.com/compute/v1/projects/some-project-id/global/networks/vertex-ai-manual-test-vpc-v2.
  | To change the VPC network, please undeploy all Vertex AI deployment resources, delete all endpoint resources, and then retry creating resources in 30 mins.
```

File: mmv1/third_party/terraform/tests/resource_vertex_ai_index_endpoint_test.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+  name       = "%{network_name}"
+}
+data "google_project" "project" {}
+`, context)
+}
+
+func testAccVertexAIIndexEndpoint_updated(context map[string]interface{}) string {
+	return Nprintf(`
+resource "google_vertex_ai_index_endpoint" "index_endpoint" {
+  display_name = "sample-endpoint"
```
[the remark this answers is not in the document]
Comment (shotarok) [reply mmv1/third_party/terraform/tests/resource_vertex_ai_index_endpoint_test.go:82]: I could confirm the display_name was immutable by running the test locally. I'll update the test case. Thanks!

**Comment by shotarok** [comment]
@roaks3 I updated the code based on your feedback. Please take a look when you have time. Thanks!

**Comment by roaks3** [comment]
Awesome, this is ready to go!

I checked the CLA issue, and it does appear you'll need to rebase to remove my `gmail.com`-authored commit (sorry for the inconvenience):
1. Use `git rebase -i`.
2. Type **e** or **edit** on the commits you want to update.
3. For each commit, run: `git commit --amend --reset-author` and `git rebase --continue`
4. A force push to remote branch is required if you have pushed your branch before.

**Comment by shotarok** [comment]
@roaks3 Yay, I completed the git-rebase. Could you please approve the workflow again?

**Comment by roaks3** [comment]
@slevenick this is now blocked on your prior review where changes were requested. It looks good to me, so would you mind clearing your review so we can merge, or giving it another pass?

**Comment by shotarok** [comment]
Hello @roaks3, It looks like @slevenick unblocked this PR to merge by removing the past change request. Can we merge this PR with your approval to wrap up?

**Comment by roaks3** [comment]
Thanks for the ping! Yep, we can go ahead and merge. I'm going to wait just a bit as GitHub is having some issues, but once things look stable we can move forward.

**Comment by shotarok** [comment]
Thank you all for your kind support and feedback!

**Pull request merged**